File: C++/Item.cpp
#include "Item.h"

Item::Item(std::string name) {
	Name = name;
	Symbol = Name[0];
}

Item::Item() {
    Name = "";
    Symbol = 0;
}

std::string Item::getName() const  {
	return Name;
}

char Item::getSymbol() {
	return Symbol;
}

Weapon::Weapon(std::string name, int damage) : Item(name), damage(damage) {
    enhancement = 0;
}

Weapon::Weapon(std::string name, int damage, int enhancement) : Item(name), damage(damage), enhancement(enhancement) {}

Weapon::Weapon() {
    Name = "";
    damage = 0;
    enhancement = 0;
}

Weapon Weapon::operator+(const Weapon& other) {
    if (this->Name == other.Name) {
        return Weapon(this->Name, this->damage, this->enhancement + 1);
    }
    return *this;
}

Weapon& Weapon::operator++() {
    this->enhancement++;
    return *this;
}

Weapon Weapon::operator++(int) {
    Weapon temp = *this;
    this->enhancement++;
    return temp;
}

int Weapon::getDamage() {
    return damage;
}

std::string Weapon::getName() const {
    return Name + " +" + std::to_string(enhancement);
}

File: Java/Item.java
package Java;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Item implements Cloneable {
    protected String name;
    protected char symbol;
    protected List<String> tags;

    public Item(String name) {
        this.name = name;
        this.symbol = name.charAt(0);
        this.tags = new ArrayList<>();
    }

    public Item() {
        this.name = "";
        this.symbol = '\0';
        this.tags = new ArrayList<>();
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public char getSymbol() {
        return symbol;
    }

    public void addTag(String tag) {
        tags.add(tag);
    }

     // Мелкое клонирование
    @Override
    protected Object clone() throws CloneNotSupportedException {
        return super.clone();
    }

    // Глубокое клонирование
    public Item deepClone() throws CloneNotSupportedException {
        Item cloned = (Item) super.clone();
        cloned.tags = new ArrayList<>(this.tags);
        return cloned;
    }

    @Override
    public String toString() {
        return "Item name: " + this.name + " " + tags.toString();
    }
}
